### PR TITLE
feat: path utils

### PR DESCRIPTION
**Pull request by P0lip**
Just a handful of them.
LMK if we need more and in what form.

### CODE REVIEW

File: src/path.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import * as path from 'path';
```
Comment (marbemac) [inline src/path.ts]: For the tests, could we perhaps run tests twice, once with path mocked/set to `path.posix` and once with `path.win32`? https://nodejs.org/api/path.html#path_windows_vs_posix

File: src/path.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+import * as URI from 'urijs';
+import { URI as VSURI } from 'vscode-uri';
+
+export const isAbsolute = (filepath: string) => path.isAbsolute(filepath) || new URI(filepath).is('absolute');
+
+export const join = (...parts: string[]) => parts.join('/');
```
Comment (marbemac) [inline src/path.ts]: `path.normalize(filepath).replace(/\\/g, '/');`?

File: src/path.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+export const normalize = (filepath: string) => path.normalize(filepath).replace(/\\/g, '/');
+
+export function toFSPath(uri: string) {
+  return path.normalize(VSURI.file(uri).fsPath);
+}
```
Comment (marbemac) [inline src/path.ts]: Can we add a relative helper? https://nodejs.org/api/path.html#path_path_relative_from_to

File: src/path.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import * as path from 'path';
+import * as URI from 'urijs';
+import { URI as VSURI } from 'vscode-uri';
+
+export const isAbsolute = (filepath: string) => path.isAbsolute(filepath) || new URI(filepath).is('absolute');
+
+export const join = (...parts: string[]) => parts.join('/');
+export const normalize = (filepath: string) => path.normalize(filepath).replace(/\\/g, '/');
```
Comment (marbemac) [inline src/path.ts]: What happens when non file e.g. `https://localhost:3000/bar.json?page=1#hello` is pass into all of these helpers? Can you add a few test cases for this?

**Comment by marbemac** [comment]
Another question - what causes these methods to throw an exception, if anything? Might be useful to at least understand that via some tests. I know vscode-uri throws an error if there is no authority set, so that toFSPath function can definitely throw, so consumers should be careful.

**Comment by wmhilton** [comment]
I don't understand why we need any of them. 🤷‍♂ 

Also, I strongly feel that using the `path` module is simply a footgun. It's a piece of shit. I've read the source code. I've read the source for `path-browserify`. Both the current version and the old outdated version that's included with Webpack 4. The `path` from Node and the `path` in Webpack 4 have major breaking differences (notably `path.posix` and `path.windows` is completely missing from the Webpack version).

The code of `path` is [such a mess](https://github.com/nodejs/node/blob/master/lib/path.js), I think it has passed the limits of human ability to reason about it. We'll never be able to write predictable isomorphic code using it.

We should just stick to using simple string functions, manipulating a normalized subset of file path syntax we can reasonably understand, i.e. our forward-slash normalized lower-case-drive-letter form.

File: src/path.ts (hunk 1 above)
Comment (wmhilton) [reply src/path.ts]: no. Just don't use `path`

**Comment by wmhilton** [comment]
Why are we putting these in `json-ref-resolver` anyway? How will we re-use these across graphite and studio?

I guess at this point I'm basically arguing we need a `@stoplight/path` module.

**Comment by wmhilton** [comment]
There's a number of useful path utils in isomorphic-git
- [normalizePath](https://github.com/isomorphic-git/isomorphic-git/blob/master/src/utils/normalizePath.js)
- [join](https://github.com/isomorphic-git/isomorphic-git/blob/master/src/utils/join.js)
- [basename](https://github.com/isomorphic-git/isomorphic-git/blob/master/src/utils/basename.js)
- [dirname](https://github.com/isomorphic-git/isomorphic-git/blob/master/src/utils/dirname.js)

**Comment by P0lip** [comment]
> I guess at this point I'm basically arguing we need a @stoplight/path module.

Yeah, we could have a pkg. This is something I suggested Marc before I created this PR.
I'm okay with switching to plain string manipulation

**Comment by marbemac** [comment]
Fineeee - `@stoplight/path` or `@stoplight/uri`?

Re usage of `path` module - what is the safest option? If us re-building parts of the `path` module is safest, then 👍, if leveraging the `path` module is safest then I'd suggest we go with that even if its code quality is suspect.

Keep in mind our "special" use case is that we always want forward slashes.

@wmhilton an important feature that is non trivial is the handling of directory traversing that path.normalize/join takes care of (`c:/foo/../bar.json`) - do the isomorphic utils handle this?

**Comment by wmhilton** [comment]
> Fineeee - `@stoplight/path` or` @stoplight/uri`?

Since these aren't URIs I would definitely prefer `@stoplight/path` :)

> what is the safest option?

Build it ourself. Ideally with a proper parser. I've already started looking at parser-generators... they may be overkill but it gives me something to prototype with.

> an important feature that is non trivial is the handling of directory traversing that path.normalize/join takes care of (c:/foo/../bar.json) - do the isomorphic utils handle this?

I haven't written one that does this _yet_ but plan to 👍

**Comment by marbemac** [comment]
> Since these aren't URIs I would definitely prefer @stoplight/path :)

Just keep in mind we should support URIs AND file paths :). The primary URI use case being URLs right now.

@P0lip will open up a PR for the separate package tomorrow.

**Comment by P0lip** [comment]
> Build it ourself. Ideally with a proper parser. I've already started looking at parser-generators... they may be overkill but it gives me something to prototype with.

jison might be quite okay.

**Pull request closed without merging**